src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -3116,13 +3112,13 @@ void G1CollectedHeap::do_collection_pause_at_safepoint_helper(double target_paus
   // without its logging output interfering with the logging output
   // that came from the pause.
 
-  if (should_start_conc_mark) {
-    // CAUTION: after the doConcurrentMark() call below, the concurrent marking
+  if (should_start_concurrent_mark_operation) {
+    // CAUTION: after the start_concurrent_mark() call below, the concurrent marking
     // thread(s) could be running concurrently with us. Make sure that anything
     // after this point does not assume that we are the only GC thread running.
     // Note: of course, the actual marking work will not start until the safepoint
     // itself is released in SuspendibleThreadSet::desynchronize().
-    do_concurrent_mark();
+    start_concurrent_cycle(concurrent_operation_is_mark_cycle /* full_mark */);
```

Comment:
Instead of passing the state down here what do you think of something like:
```suggestion
    if (_collector_state->cancel_conc_start()) {
      cancel_concurrent_cycle();
    } else {
      start_concurrent_cycle();
    }
```

Where `cancel_concurrent_cycle()` would do:
```
  root_regions()->cancel_scan();
  MutexLocker x(CGC_lock, Mutex::_no_safepoint_check_flag);
  if (!_cm_thread->in_progress()) {
    _cm_thread->set_canceled();
    CGC_lock->notify();
  }
```